### PR TITLE
Implement target gzip request compression

### PR DESCRIPTION
**Pull request by AndreKurait**
### Description

Implement optional target gzip request compression.
This can reduce bottlenecks on the networking layer from the RFS workers.

* Category: New Feature
* Why these changes are required?
* What is the old behavior before changes and new behavior after changes?

### Issues Resolved
[MIGRATIONS-1964](https://opensearch.atlassian.net/browse/MIGRATIONS-1964)

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
Docker integ-type tests added as well as unit tests.

Tested in AWS, You can see in the image the decrease in "bytesSent" and overall more consistent throughput.
More testing needs to be done to identify the cases where performance is increased.

Testing, 1 task, no target cluster throttling:  "cpu": "4096", "memory": "8192", "platform": "ARM"
Left, no compression. Right, compression
image

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: RFS/src/main/java/com/rfs/common/http/GzipRequestTransformer.java
Comment (peternied) [inline RFS/src/main/java/com/rfs/common/http/GzipRequestTransformer.java:21]: I'm not sure why we are using a separate transformer for this. `HttpClient.create()` supports `.compress(true/false)` that creates a client that uses compression, can we use that instead of building our own?

Comment (AndreKurait) [reply RFS/src/main/java/com/rfs/common/http/GzipRequestTransformer.java:21]: That's to request/handle response compression, not request.

The main reason to compress is to reduce bandwidth used between RFS and Target. From the screenshot in the overview, requests can account for 75% of the total bandwidth so there is more opportunity to reduce bottlenecks with request compression.

There may be some value in enabling response compression by default too, it's less likely to have unintended consequences since the server can just ignore the request if it's not supported.

Comment (peternied) [reply RFS/src/main/java/com/rfs/common/http/GzipRequestTransformer.java:21]: Thanks for the details

File: DocumentsFromSnapshotMigration/README.md
Comment (peternied) [reply DocumentsFromSnapshotMigration/README.md:74]: Or alternatively, what are we missing to commit to always supporting compressed responses?

Comment (AndreKurait) [reply DocumentsFromSnapshotMigration/README.md:74]: `all of our supported target cluster will support compressed requests`, I know this needs to be updated with limitations, but we were declaring support for 7.10 target clusters on the [readme](https://github.com/opensearch-project/opensearch-migrations?tab=readme-ov-file#supported-source-and-target-versions)

File: RFS/src/test/java/com/rfs/common/http/GzipRequestTransformerTest.java
Comment (AndreKurait) [reply RFS/src/test/java/com/rfs/common/http/GzipRequestTransformerTest.java]: Fixed

File: RFS/src/main/java/com/rfs/common/http/GzipRequestTransformer.java
Comment (AndreKurait) [reply RFS/src/main/java/com/rfs/common/http/GzipRequestTransformer.java:1]: Fixed

File: DocumentsFromSnapshotMigration/README.md
Comment (peternied) [reply DocumentsFromSnapshotMigration/README.md:74]: Elasticsearch and OpenSearch have supported compression for a long time... I don't think we are ever going to support a pre-7.10 migration so I think this should be on by default.

Comment (peternied) [reply DocumentsFromSnapshotMigration/README.md:74]: If you feel very strongly that we cannot move forward without a cli option, then I think we need to restructure how we pass these low level parameters around in our system, I'd suggest adding a new arguments JCommander object that we can keep adding onto that includes all these 'advanced settings'.

This would be a case where I would support static references to keep our construction/callstack patterns cleaner so we don't have to pipe these int/bool values all over

Comment (AndreKurait) [reply DocumentsFromSnapshotMigration/README.md:74]: Reading [AWS Docs](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/gzip.html#:~:text=You%20can%20compress%20HTTP%20requests,with%20similar%20OpenSearch%20settings%2C%20http_compression), support is by default off on ES 6.x and on thereafter. (So i'm less concerned about target cluster support. 

On the [OpenSource](https://opensearch.org/docs/latest/install-and-configure/configuring-opensearch/network-settings/) side, it's by default OFF when HTTPS is on to mitigate security concerns.

Given it's configurable, i'd rather leave a cli option in. Does your thoughts on this being an "advanced option" change in light of this?

Comment (gregschohn) [reply DocumentsFromSnapshotMigration/README.md:74]: Are we sure universally that request compression is always a net gain for the client and the server?  I suspect that there will be a number of workloads where it will NOT be.  Until we're sure, having a flag to toggle this seems like a prudent hedge.  Compression also makes debugging considerably more difficult (I usually turn it off in both directions when trying to debug)

File: DocumentsFromSnapshotMigration/README.md
Comment (gregschohn) [inline DocumentsFromSnapshotMigration/README.md:77]: why did this value change?

File: DocumentsFromSnapshotMigration/src/test/java/com/rfs/ParallelDocumentMigrationsTest.java
Comment (gregschohn) [inline DocumentsFromSnapshotMigration/src/test/java/com/rfs/ParallelDocumentMigrationsTest.java:55]: I don't think that you're splicing these values into the return value.

File: RFS/src/main/java/com/rfs/common/http/GzipRequestTransformer.java
Comment (gregschohn) [inline RFS/src/main/java/com/rfs/common/http/GzipRequestTransformer.java:49]: two questions - why not just use the simple single-argument constructor?  Why do you have a subclass of FastGzipOutputStream (setting the level in this method)?

**Review by gregschohn — COMMENTED** [review]
One question on the parameterized test - other than that, just nits and general information requests.

File: DocumentsFromSnapshotMigration/README.md
Comment (AndreKurait) [reply DocumentsFromSnapshotMigration/README.md:77]: Changed outside of this PR, was reconciling the actual value with the docs here.

When I had been testing with 4GB workers and 10MB requests on ECS, I got better throughput with 10 connections than 20.

We can and should do further testing to identify good starting out values for this.

File: DocumentsFromSnapshotMigration/README.md
Comment (peternied) [reply DocumentsFromSnapshotMigration/README.md:74]: I think we need to restructure how we pass these low level parameters around in our system, I'd suggest adding a new arguments JCommander object that we can keep adding onto that includes all these 'advanced settings'.

This would be a case where I would support static references to keep our construction/callstack patterns cleaner so we don't have to pipe these int/bool values all over

File: RFS/src/main/java/com/rfs/common/http/GzipRequestTransformer.java
Comment (AndreKurait) [reply RFS/src/main/java/com/rfs/common/http/GzipRequestTransformer.java:49]: The Deflator in DeflaterOutputStream (extended in GZIPOutputStream) is protected so i cannot set the level without  the subclass

File: DocumentsFromSnapshotMigration/src/test/java/com/rfs/ParallelDocumentMigrationsTest.java
Comment (AndreKurait) [reply DocumentsFromSnapshotMigration/src/test/java/com/rfs/ParallelDocumentMigrationsTest.java:53]: Fixed

File: DocumentsFromSnapshotMigration/README.md
Comment (AndreKurait) [reply DocumentsFromSnapshotMigration/README.md:74]: Refactored into a TargetAdvancedArgs object. @peternied, let me know your thoughts

Comment (peternied) [reply DocumentsFromSnapshotMigration/README.md]: Thanks for the update there.  I like the direction you took this, and I've created a jira for some additional follow up that might make sense to iterate on if we find ourselves injecting new parameters https://opensearch.atlassian.net/browse/MIGRATIONS-1972

**Comment by peternied** [comment]
All my outstanding comments have been addressed @gregschohn I'm not sure if all of your concerns have been addressed - I'll defer to your approval.

**Review by gregschohn — APPROVED** [review]

File: RFS/src/test/java/com/rfs/common/http/GzipPayloadRequestTransformerTest.java
Comment (AndreKurait) [reply RFS/src/test/java/com/rfs/common/http/GzipPayloadRequestTransformerTest.java:254]: We don't regularly use limits so i could remove this, this is just helping to check that the buffer is duplicated before being read

File: RFS/src/main/java/com/rfs/common/http/GzipPayloadRequestTransformer.java
Comment (AndreKurait) [reply RFS/src/main/java/com/rfs/common/http/GzipPayloadRequestTransformer.java:71]: In the current code, we don't have any places we would expect multiple passes through this transformer so it could be an assertion.

**Pull request merged**
